common/utils/src/main/scala/org/apache/spark/internal/README.md

```diff
@@ -45,3 +45,52 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
 ## Exceptions
 
 To ensure logs are compatible with Spark SQL and log analysis tools, avoid `Exception.printStackTrace()`. Use `logError`, `logWarning`, and `logInfo` methods from the `Logging` trait to log exceptions, maintaining structured and parsable logs.
+
+## External third-party ecosystem access
+
+* If you want to output logs in `scala code` through the structured log framework, you can define `custom LogKeys` as follows:
+
+```scala
+object CustomLogKeys {
+  // External third-party ecosystem `custom LogKey` must be `extends LogKey`
+  case object CUSTOM_LOG_KEY extends LogKey
+}
+```
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;and use it in `scala` code:
+
+```scala
+logInfo(log"${MDC(CustomLogKeys.CUSTOM_LOG_KEY, "key")}")
+```
+
+* If you want to output logs in `java code` through the structured log framework, you can define `custom LogKeys` as follows:
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;when your project does **not depend on `scala`**,
+```java
+class CustomLogKeys {
+  // External third-party ecosystem `custom LogKey` must be `implements LogKey`
+  public static class CUSTOM_LOG_KEY implements LogKey { }
+
+  // Singleton
+  public static final CUSTOM_LOG_KEY CUSTOM_LOG_KEY = new CUSTOM_LOG_KEY();
+}
+```
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;and use it in `java` code:
```

Comment:
why do we have &nbsp here?

Reply:
If `&nbsp` is not added, its `indent` looks incorrect, as shown below:
<img width="1131" alt="image" src="https://github.com/apache/spark/assets/15246973/18474be8-ada9-456e-bc08-decb46b74e3c">

Let me incorporate this sentence into the above one, and revise it as follows:
<img width="1133" alt="image" src="https://github.com/apache/spark/assets/15246973/45e289ad-e948-4f4f-8789-173c156b7c44">